### PR TITLE
[TOPIC-BLE-LLCP]: Bluetooth: llcp: release ctx at end of procedure

### PR DESCRIPTION
**Pull request by kruithofa**
When a control procedure is finished, or upon disconnection, we need
to release the context to avoid running out of memory

Signed-off-by: Andries Kruithof <Andries.Kruithof@nordicsemi.no>

### CODE REVIEW

File: subsys/bluetooth/controller/ll_sw/ull_llcp.c [hunk 1]
```diff
@@ -169,7 +172,9 @@ static struct proc_ctx *create_procedure(enum llcp_proc proc)
 	/* Initialize opcodes fields to  known values */
 	ctx->rx_opcode = ULL_LLCP_INVALID_OPCODE;
 	ctx->tx_opcode = ULL_LLCP_INVALID_OPCODE;
-	//ctx->response_opcode = 0xFFU;
+	ctx->response_opcode = INITIAL_VALUE;
```
Comment (thoh-ot) [inline subsys/bluetooth/controller/ll_sw/ull_llcp.c:175]: Use ULL_LLCP_INVALID_OPCODE

File: subsys/bluetooth/controller/ll_sw/ull_llcp.c [hunk 2]
```diff
@@ -169,7 +172,9 @@ static struct proc_ctx *create_procedure(enum llcp_proc proc)
...
+
+	ctx->release_ctx = 0;
```
Comment (thoh-ot) [inline subsys/bluetooth/controller/ll_sw/ull_llcp.c:177]: Move this up to the other flags and use 0U

File: subsys/bluetooth/controller/ll_sw/ull_llcp_local.c [hunk 3]
```diff
@@ -173,12 +175,23 @@ static void lr_act_run(struct ll_conn *conn)
 		/* Unknown procedure */
 		LL_ASSERT(0);
 	}
+
+	LL_ASSERT(ctx != NULL);
```
Comment (thoh-ot) [inline subsys/bluetooth/controller/ll_sw/ull_llcp_local.c:179]: Nothing reassigns ctx so NULL check is pointless

File: subsys/bluetooth/controller/ll_sw/ull_llcp.c [hunk 4]
```diff
@@ -109,6 +110,7 @@ struct node_tx *ull_cp_priv_tx_alloc(void)
 
 static void tx_release(struct node_tx *tx)
 {
+	memset(tx, INITIAL_VALUE, sizeof(struct node_tx));
```
Comment (thoh-ot) [inline subsys/bluetooth/controller/ll_sw/ull_llcp.c:113]: Seems unrelated to this PR

File: subsys/bluetooth/controller/ll_sw/ull_llcp.c [hunk 5]
```diff
@@ -127,6 +129,7 @@ struct node_rx_pdu *ull_cp_priv_ntf_alloc(void)
 
 static void ntf_release(struct node_rx_pdu *ntf)
 {
+	memset(ntf, INITIAL_VALUE, sizeof(struct node_rx_pdu));
```
Comment (thoh-ot) [inline subsys/bluetooth/controller/ll_sw/ull_llcp.c:132]: Seems unrelated to this PR

File: subsys/bluetooth/controller/ll_sw/ull_llcp_local.c [hunk 6]
```diff
@@ -140,6 +140,8 @@ static void lr_act_run(struct ll_conn *conn)
 	struct proc_ctx *ctx;
 
 	ctx = lr_peek(conn);
+	LL_ASSERT(ctx != NULL);
```
Comment (thoh-ot) [inline subsys/bluetooth/controller/ll_sw/ull_llcp_local.c:143]: lr_act_run is already guarded at callsite to have a valid context at the head of the queue

File: subsys/bluetooth/controller/ll_sw/ull_llcp_local.c [hunk 7]
```diff
@@ -173,12 +175,23 @@ static void lr_act_run(struct ll_conn *conn)
 		/* Unknown procedure */
 		LL_ASSERT(0);
 	}
+
+	LL_ASSERT(ctx != NULL);
+	if ((ctx != NULL) && ctx->release_ctx) {
+		ctx = lr_dequeue(conn);
+		LL_ASSERT(ctx != NULL);
```
Comment (thoh-ot) [inline subsys/bluetooth/controller/ll_sw/ull_llcp_local.c:182]: Do a peek into the queue and compare it with the ctx instead

File: subsys/bluetooth/controller/ll_sw/ull_llcp_remote.c [hunk 8]
```diff
@@ -197,6 +197,8 @@ static void rr_act_run(struct ll_conn *conn)
 	struct proc_ctx *ctx;
 
 	ctx = rr_peek(conn);
+	LL_ASSERT(ctx != NULL);
```
Comment (thoh-ot) [inline subsys/bluetooth/controller/ll_sw/ull_llcp_remote.c:200]: rr_act_run is already guarded at callsite to have a valid context at the head of the queue

File: subsys/bluetooth/controller/ll_sw/ull_llcp_remote.c [hunk 9]
```diff
@@ -230,6 +232,14 @@ static void rr_act_run(struct ll_conn *conn)
 		/* Unknown procedure */
 		LL_ASSERT(0);
 	}
+
+
+	LL_ASSERT(ctx != NULL);
```
Comment (thoh-ot) [inline subsys/bluetooth/controller/ll_sw/ull_llcp_remote.c:237]: Nothing reassigns ctx so NULL check is pointless

File: subsys/bluetooth/controller/ll_sw/ull_llcp_remote.c [hunk 10]
```diff
@@ -230,6 +232,14 @@ static void rr_act_run(struct ll_conn *conn)
...
+	if (ctx->release_ctx) {
+		ctx = rr_dequeue(conn);
+		LL_ASSERT(ctx != NULL);
```
Comment (thoh-ot) [inline subsys/bluetooth/controller/ll_sw/ull_llcp_remote.c:240]: Do a peek into the queue and compare it with the ctx instead

File: subsys/bluetooth/controller/ll_sw/ull_llcp_remote.c [hunk 11]
```diff
@@ -266,21 +276,25 @@ static void rr_act_reject(struct ll_conn *conn)
 	} else {
 		struct proc_ctx *ctx = rr_peek(conn);
 
+		LL_ASSERT(ctx != NULL);
 		rr_tx(conn, ctx, PDU_DATA_LLCTRL_TYPE_REJECT_IND);
 
-		/* Dequeue pending request that just completed */
-		(void) rr_dequeue(conn);
-
+		/* The pending request is dequeued later */
```
Comment (thoh-ot) [inline subsys/bluetooth/controller/ll_sw/ull_llcp_remote.c:282]: Missing ctx->release_ctx = 1U

File: subsys/bluetooth/controller/ll_sw/ull_llcp_remote.c [hunk 12]
```diff
@@ -266,21 +276,25 @@ static void rr_act_reject(struct ll_conn *conn)
...
+	struct proc_ctx *ctx;
+
 	rr_set_collision(conn, 0U);
 
 	/* Dequeue pending request that just completed */
-	(void) rr_dequeue(conn);
+	ctx = rr_peek(conn);
+	LL_ASSERT(ctx != NULL);
+
+	ctx->release_ctx = 1;
```
Comment (thoh-ot) [inline subsys/bluetooth/controller/ll_sw/ull_llcp_remote.c:297]: Use 1U

File: subsys/bluetooth/controller/ll_sw/ull_llcp_remote.c [hunk 13]
```diff
@@ -310,6 +324,14 @@ static void rr_st_idle(struct ll_conn *conn, uint8_t evt, void *param)
...
 	struct proc_ctx *ctx;
 
+	/*
+	 * check if we need to release the context
+	 */
+	ctx = rr_peek(conn);
+	if (ctx != NULL && ctx->release_ctx) {
+		ctx = rr_dequeue(conn);
+		proc_ctx_release(ctx);
+	}
```
Comment (thoh-ot) [inline subsys/bluetooth/controller/ll_sw/ull_llcp_remote.c:334]: What does this solve?

File: subsys/bluetooth/controller/ll_sw/ull_llcp_local.c [hunk 14]
```diff
@@ -173,12 +175,23 @@ static void lr_act_run(struct ll_conn *conn)
...
 static void lr_act_complete(struct ll_conn *conn)
 {
-	/* Dequeue pending request that just completed */
-	(void) lr_dequeue(conn);
+	struct proc_ctx *ctx;
+
+	ctx = lr_peek(conn);
+	LL_ASSERT(ctx != NULL);
+
+	ctx->release_ctx = 1;
```
Comment (thoh-ot) [inline subsys/bluetooth/controller/ll_sw/ull_llcp_local.c:194]: Use 1U

**Review by thoh-ot — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/controller/ll_sw/ull_llcp_local.c [hunk 15]
```diff
@@ -206,6 +231,14 @@ static void lr_st_disconnect(struct ll_conn *conn, uint8_t evt, void *param)
...
 static void lr_st_idle(struct ll_conn *conn, uint8_t evt, void *param)
 {
+	struct proc_ctx *ctx;
+
+	ctx = lr_peek(conn);
+	if ((ctx != NULL) && (ctx->release_ctx)) {
+		ctx = lr_dequeue(conn);
+		proc_ctx_release(ctx);
+	}
+
```
[the remark this answers is not in the document]
Comment (kruithofa) [reply subsys/bluetooth/controller/ll_sw/ull_llcp_local.c:241]: Removed since by this time the context should already be released

**Review by erbr-ot — APPROVED** [review]
LFTM - although I agree with THOH to loose the llcp in llcp_done and related funcs

**Review by thoh-ot — APPROVED** [review]
